### ISSUE TITLE
Things aren't deterministic across platforms

### ISSUE THREAD

**Issue by Daylily-Zeleen** [godot-jolt/godot-jolt#301]
I desire a cross platiform deterministic solution for  physics, and noticed this repo.

The JoltPhysics support cross platform deterministic, I don't the principle behind that, I am not an expert of computer and compiler.

I noticed that the JoltPhysics set floating-point behavior to `precise` and turn  off `contract`, and disable  `FMADD` to realize that.

I am not familiar with cmake, it seem the `CMakeLists.txt` alaway set floating-point behavior to be `fast`.

Can we do some effort to realize cross platform deterministic base on this repo?

If it is very hard, can you give me some suggestions about this topic?

**Comment by mihe** [comment]
I haven't documented this yet, but this project will not be making any guarantees about determinism, and will explicitly not support cross-platform determinism, regardless of what Jolt itself supports.

There is a project setting called `physics/jolt_3d/simulation/more_deterministic` that enables Jolt's deterministic simulation, but while this makes Jolt itself deterministic (when using the same binary) there could still be things within this extension or within Godot itself that interferes with this determinism. That's why the project setting is named "More Deterministic" and not "Deterministic". I realize determinism is not all that useful unless it's a guarantee, but I figured someone might appreciate slightly more deterministic behavior, hence the project setting.

I used to provide a CMake option for cross-platform determinism, mostly just to align with Jolt's compiler flags, but I removed it in pull request #235. This had never actually been tested, so I have no idea if it worked. I would assume that it did not.

I'm not an expert on determinism, but as I understand it the entire "pipeline" from the user's scripts in Godot to something like (for example) the `AddForce` call on my end needs to take determinism into account. That "pipeline" obviously includes the Godot engine itself, which means you now have to compile Godot from source with the same strict compile flags and somehow verify that this actually results in deterministic behavior.

Even if it was possible to compile everything with these compile flags and somehow verify that the entire "pipeline" is indeed deterministic, it would be a constant battle so long as Godot itself doesn't make the same guarantees about determinism, since any pull request to Godot could break this.

I believe the only reliable way of achieving cross-platform determinism in Godot is to do something like what [SG Physics 2D](https://gitlab.com/snopek-games/sg-physics-2d) does, which uses fixed-point math instead of floating-point math, and also doesn't implement the proper `PhysicsServer2D`, which means the user has much better control over how and when the physics simulation is interacted with.
